File: android/app/src/main/kotlin/com/truemap/app/ReconstructionViewModel.kt
package com.truemap.app

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.launch

class ReconstructionViewModel : ViewModel() {
    var videoUrls by mutableStateOf<List<String>>(emptyList())
        private set
    
    var isProcessing by mutableStateOf(false)
        private set
    
    var progress by mutableStateOf(0.0)
        private set
    
    var statusMessage by mutableStateOf("")
        private set
    
    var hasCompletedReconstruction by mutableStateOf(false)
        private set
    
    var errorMessage by mutableStateOf<String?>(null)
        private set
    
    private val reconstructionService = ReconstructionService()
    
    init {
        // Observe reconstruction service updates
        reconstructionService.setProgressCallback { status, prog, message ->
            viewModelScope.launch {
                progress = prog
                statusMessage = message
                isProcessing = (status != ProcessingStatus.IDLE && 
                               status != ProcessingStatus.COMPLETED && 
                               status != ProcessingStatus.ERROR)
                hasCompletedReconstruction = (status == ProcessingStatus.COMPLETED)
            }
        }
    }
    
    fun addVideoUrl(url: String) {
        if (url.isNotEmpty()) {
            videoUrls = videoUrls + url
        }
    }
    
    fun removeUrl(url: String) {
        videoUrls = videoUrls - url
    }
    
    fun startReconstruction() {
        if (videoUrls.isEmpty()) {
            errorMessage = "Please add at least one video URL"
            return
        }
        
        isProcessing = true
        hasCompletedReconstruction = false
        progress = 0.0
        statusMessage = "Starting reconstruction..."
        
        viewModelScope.launch {
            try {
                reconstructionService.startReconstruction(videoUrls)
                statusMessage = "Reconstruction completed successfully!"
            } catch (e: Exception) {
                errorMessage = "Reconstruction failed: ${e.message}"
                isProcessing = false
            }
        }
    }
    
    fun cancelReconstruction() {
        reconstructionService.cancelReconstruction()
        isProcessing = false
        statusMessage = "Cancelled"
    }
    
    fun exportMesh() {
        viewModelScope.launch {
            try {
                val path = reconstructionService.exportMesh()
                statusMessage = "Mesh exported to: $path"
            } catch (e: Exception) {
                errorMessage = "Export failed: ${e.message}"
            }
        }
    }
    
    fun clearError() {
        errorMessage = null
    }
}

File: ios/Truemap/ViewModels/ReconstructionViewModel.swift
import Foundation
import Combine

/// ViewModel for managing 3D reconstruction process
class ReconstructionViewModel: ObservableObject {
    @Published var videoURLs: [String] = []
    @Published var isProcessing = false
    @Published var progress: Double = 0.0
    @Published var statusMessage = ""
    @Published var hasCompletedReconstruction = false
    @Published var errorMessage: String?
    
    private let reconstructionService = ReconstructionService()
    private var cancellables = Set<AnyCancellable>()
    
    init() {
        // Subscribe to reconstruction service updates
        reconstructionService.progressPublisher
            .receive(on: DispatchQueue.main)
            .sink { [weak self] progressUpdate in
                self?.progress = progressUpdate.progress
                self?.statusMessage = progressUpdate.message
            }
            .store(in: &cancellables)
        
        reconstructionService.statusPublisher
            .receive(on: DispatchQueue.main)
            .sink { [weak self] status in
                self?.isProcessing = (status != .idle && status != .completed && status != .error)
                self?.hasCompletedReconstruction = (status == .completed)
            }
            .store(in: &cancellables)
    }
    
    func addVideoURL(_ url: String) {
        guard !url.isEmpty else { return }
        videoURLs.append(url)
    }
    
    func deleteURL(at offsets: IndexSet) {
        videoURLs.remove(atOffsets: offsets)
    }
    
    func startReconstruction() {
        guard !videoURLs.isEmpty else {
            errorMessage = "Please add at least one video URL"
            return
        }
        
        isProcessing = true
        hasCompletedReconstruction = false
        progress = 0.0
        statusMessage = "Starting reconstruction..."
        
        reconstructionService.startReconstruction(urls: videoURLs) { [weak self] result in
            DispatchQueue.main.async {
                switch result {
                case .success:
                    self?.statusMessage = "Reconstruction completed successfully!"
                case .failure(let error):
                    self?.errorMessage = error.localizedDescription
                    self?.isProcessing = false
                }
            }
        }
    }
    
    func cancelReconstruction() {
        reconstructionService.cancelReconstruction()
        isProcessing = false
        statusMessage = "Cancelled"
    }
    
    func exportMesh() {
        reconstructionService.exportMesh { [weak self] result in
            DispatchQueue.main.async {
                switch result {
                case .success(let url):
                    self?.statusMessage = "Mesh exported to: \(url.lastPathComponent)"
                    // In a real app, would present share sheet here
                case .failure(let error):
                    self?.errorMessage = error.localizedDescription
                }
            }
        }
    }
}
